### PR TITLE
add python inference api

### PR DESCRIPTION
**Pull request by fc500110**
Python inference API, support `AnalysisPredictor` and `NativePredictor`

### CODE REVIEW

File: python/paddle/fluid/executor.py [hunk 1]
```diff
@@ -452,7 +454,9 @@ def run(self,
             fetch_var_name='fetch',
             scope=None,
             return_numpy=True,
-            use_program_cache=False):
+            use_program_cache=False,
+            infer_inputs=None,
```
Comment (panyx0718) [inline python/paddle/fluid/executor.py]: reuse `feed` and behave the same?

File: python/paddle/fluid/executor.py [hunk 2]
```diff
@@ -452,7 +454,9 @@ def run(self,
...
+            infer_attrs=None):
```
Comment (panyx0718) [inline python/paddle/fluid/executor.py]: what is this for? If this is config, it should be put in with_inference_optimize.

File: python/paddle/fluid/executor.py [hunk 3]
```diff
@@ -522,6 +528,15 @@ def run(self,
                 scope=scope,
                 return_numpy=return_numpy,
                 use_program_cache=use_program_cache)
+        else:
```
Comment (panyx0718) [inline python/paddle/fluid/executor.py]: take a look at L543, and do something similar.

File: python/paddle/fluid/executor.py [hunk 4]
```diff
@@ -27,6 +27,8 @@
 __all__ = ['Executor', 'global_scope', 'scope_guard']
 
 g_scope = core.Scope()
+InferNativeConfig = core.NativeConfig
+InferAnalysisConfig = core.AnalysisConfig
```
Comment (panyx0718) [inline python/paddle/fluid/executor.py:31]: What do you need 2 Config? Can it be done in one config?

File: python/paddle/fluid/compiler.py [hunk 5]
```diff
@@ -107,11 +109,18 @@ def with_data_parallel(self,
...
+            isinstance(config, InferAnalysisConfig)
+        ])
+        self._infer_config = config
+
     def _with_distributed(self):
         raise NotImplementedError()
 
     def _with_inference_optimize(self):
-        raise NotImplementedError()
+        self._infer_predictor = core.create_paddle_predictor(self._infer_config)
```
Comment (panyx0718) [inline python/paddle/fluid/compiler.py]: Predictor is created in _compile

**Review by panyx0718 — COMMENTED** [review]
Please follow the example of data_parallel and be as consistent as possible.

File: python/paddle/fluid/executor.py (hunk 2 above)
Comment (fc500110) [reply python/paddle/fluid/executor.py]: config is used to create paddle predictor, `infer_attrs` is inference attribute, etc. `batch_size`, `wormup`.

File: python/paddle/fluid/executor.py (hunk 4 above)
Comment (Superjomn) [reply python/paddle/fluid/executor.py:31]: The `AnalysisConfig` and `NativeConfig` are exposed from C++ inference predictor directly, we hope the python API works as similar to the c++ inference as possible, so the existing inference users will be more clear facing the new python APIs.

The `AnalysisConfig` will create AnalysisPredictor while the NativeConfig will create NativePredictor, that is a quite common concept in the C++ inference.

Comment (fc500110) [reply python/paddle/fluid/executor.py:31]: The two `Config` class correspond to `NativePredictor` and `AnalysisPredictor`. There is no class inheritance relations.

File: python/paddle/fluid/compiler.py (hunk 5 above)
Comment (fc500110) [reply python/paddle/fluid/compiler.py]: fixed.

File: python/paddle/fluid/executor.py (hunk 3 above)
Comment (fc500110) [reply python/paddle/fluid/executor.py]: Could you tell me more details？

File: python/paddle/fluid/executor.py (hunk 1 above)
Comment (fc500110) [reply python/paddle/fluid/executor.py]: fixed.

File: python/paddle/fluid/executor.py (hunk 3 above)
Comment (fc500110) [reply python/paddle/fluid/executor.py]: The code has been refactored.

File: python/paddle/fluid/executor.py (hunk 2 above)
Comment (panyx0718) [reply python/paddle/fluid/executor.py]: warmup is config? batch size can be found in `feed`?

File: python/paddle/fluid/executor.py (hunk 4 above)
Comment (panyx0718) [reply python/paddle/fluid/executor.py:31]: Then they should be wrapped in one config?

File: python/paddle/fluid/executor.py (hunk 3 above)
Comment (panyx0718) [reply python/paddle/fluid/executor.py]: Please read the executor.py can try harder to be consistent. I believe you can do it.

```
        if program._is_data_parallel:
            return self._run_parallel(...)
```

File: python/paddle/fluid/executor.py (hunk 2 above)
Comment (Superjomn) [reply python/paddle/fluid/executor.py]: warmup is not an option of the existing AnalysisConfig.

warmup is just a configuration for profiling, not for inference.

We port the inference to python just for convenient of profiling and testing, so warm up is important.

File: python/paddle/fluid/executor.py (hunk 4 above)
Comment (Superjomn) [reply python/paddle/fluid/executor.py:31]: I don't think so.

For inference, it is important is to make C++ APIs concepts consistent in python, make the python inference program consistent with the C++ version.
To wrap the existing widely used C++ program and pretend to be something else is not the first goal.

If these APIs are not suitable for the executor, we can just expose the Pybind version, make a prefix with each API like `__` to warn users not to use it.

The current `PyNativeConfig` will be something like `__NativeConfig`, and we can write some testing program for convenient.

**Review by Superjomn — DISMISSED** [review]
LGTM

File: python/paddle/fluid/compiler.py [hunk 6]
```diff
@@ -109,10 +112,17 @@ def with_data_parallel(self,
             self._build_strategy = BuildStrategy()
         return self
 
-    def _with_distributed(self):
-        raise NotImplementedError()
+    def with_inference_optimize(self, config, infer_attrs):
```
Comment (panyx0718) [inline python/paddle/fluid/compiler.py]: need doc.

**Review by panyx0718 — DISMISSED** [review]
LG API

**Review by panyx0718 — DISMISSED** [review]
LG API

**Review by panyx0718 — APPROVED** [review]

**Comment by panyx0718** [comment]
LG API

**Pull request merged**